### PR TITLE
Ignore PRIORITY frames that reference closed pushed streams

### PR DESCRIPTION
**Pull request by deweerdt**
Firefox sends a PRIORITY frame when receiving a PUSH_PROMISE frame.
Because of the racy nature of the protocol, it's very possible that the
PRIORITY frame is received once the stream is closed on H2O side.

This results in us opening a new stream that will be used to account
against `globalconf->http2.max_streams_for_priority`. Eventually, this
can lead to sending a GOAWAY frame with 'ENHANCE_YOUR_CALM'.

I believe it's safe to ignore PRIORITY frames that reference pushed
streams that are closed, since no current browser appears to use those
to reference them later.

I've tested the fix by using the following handler (and a dummy js file)
and verified that we hit the GOAWAY path without the fix, and not anymore
with the fix:

```
        mruby.handler: |
          Proc.new do |env|
            case env["PATH_INFO"]
            when "/"
              push_paths = []
              (1..100).each { |i| push_paths << "/index.js?#{i}" }
              [399, push_paths.empty? ? {} : {"link" =>
ush_paths.map{|p| "<#{p}>; rel=preload"}.join(",")}, []]
            else
              [399, {}, []]
            end
          end
```

### CODE REVIEW

**Comment by kazuho** [comment]
Thank you for the PR.

I believe that the requirements should be:
* apply the values of PRIORITY frames to pushed streams that are open
* ignore apply the values of PRIORITY frames trying to modify the priority of a closed stream

To accommodate the requirements, I think we should:
* first see if the specified stream is open, and apply the specified values if it is open
* else, ignore the specified values if the stream is already closed, regardless of whether if it is a pushed stream or a pulled stream
 * I agree that using `push_stream_ids.max_open` (or `pull_stream_ids.max_open`) is the right approach

How does it sound?

PS. I am not sure if we should memoize the specified priority for an idle pushed stream. While the spec. does not forbid such frames (what it only specfies is that _the PRIORITY frame can be sent for a stream in the "idle" or "closed" state_ (quoting [section 6.3](https://http2.github.io/http2-spec/#PRIORITY))), I cannot think of any good reason for a client having priority assigned for an pushed stream in idle state.

**Comment by deweerdt** [comment]
Thanks for your reply,

> I believe that the requirements should be:
>
>  - apply the values of PRIORITY frames to pushed streams that are open
>  - ignore apply the values of PRIORITY frames trying to modify the priority of a closed stream

I think it's not that clear cut for client issued streams. It is a bit clearer for pushed streams, because clients don't use those to build the priority tree. That's why this PR only addresses pushed streams.

Regarding client issued streams, clients do use both idle and possibly closed streams in their priority tree. I'm not sure how to fix this in the current spec, besides having two heuristics:
- one for FF style priority trees: accommodate for X maximum prioritized streams
- one for Chrome style daisy chains: keep info about the last Y closed streams

> PS. I am not sure if we should memoize the specified priority for an idle pushed stream

That's a good point actually, that also could fall in the clear cut category that we can ignore.

**Comment by kazuho** [comment]
@deweerdt Thank you for the response.

To avoid confusion, I'd like to use a chart to discuss what we should do for each case in the matrix. The chart illustrates whether if we should apply or ignore the specified priority based on the state and the initiation method of a stream.

| push-or-pull \ state | idle | reserved / open / half-closed | closed |
|---|---|---|---|
|__pull__|apply|apply|ignore\*|
|__push__|apply?\*\*|apply\*\*\*|ignore\*|

For \*, I agree that we should retain priority information of certain number of closed streams to better handle Chrome-style priority trees. But for now, 
For \*\*, I am not sure if there is any use-case (as stated in [the postscript of my comment](https://github.com/h2o/h2o/pull/1105#issuecomment-253332995)).

The issue I am worried with the proposed change is that it seems to change the behavior of \*\*\* from apply to ignore. Is that intentional?

**Comment by deweerdt** [comment]
Ah, thanks, i see now. Yes, my intent was to address the push/closed state, but i see now that i should only do so if stream is NULL. Let me update the PR.

**Comment by deweerdt** [comment]
I'm realizing that this still implies a maximum of 16 inflight pushes, i wonder if we should ignore the limit for pushed streams? After all, a client can't abuse those, since they're created server side.

**Comment by kazuho** [comment]
@deweerdt 
> I'm realizing that this still implies a maximum of 16 inflight* pushes, i wonder if we should ignore the limit for pushed streams? After all, a client can't abuse those, since they're created server side.

I cannot see such implication. My understanding is that we have a limit of 16 streams for _idle_ state with priority set, regardless of whether if it is a pushed stream or a pulled one.

That said, if you are arguing for changing the behavior of (idle, push) on the chart from _apply_ to _ignore_, I have no objection.

PS. I would also appreciate if you could look into fixing the behavior for (closed, pull). It seems to me that we are not _ignore_-ing PRIORITY frames received in such condition.

**Comment by deweerdt** [comment]
>I cannot see such implication. My understanding is that we have a limit of 16 streams for idle state with priority set, regardless of whether if it is a pushed stream or a pulled one.
Yes, that's not a problem, i got confused.

5078835 addresses the three following cases:
- closed/push
- idle/push
- closed/pull
by ignoring the priority frame. I believe no browser currently uses those later in the connection so it's safe to discard them.

File: lib/http2/connection.c [hunk 1]
```diff
@@ -614,6 +618,14 @@ static int handle_priority_frame(h2o_http2_conn_t *conn, h2o_http2_frame_t *fram
         if (h2o_http2_scheduler_get_weight(&stream->_refs.scheduler) != 257)
             set_priority(conn, stream, &payload, 1);
     } else {
+        /* Ignore PRIORITY frames for closed streams */
+        if (frame->stream_id & 1) {
```
Comment (kazuho) [inline lib/http2/connection.c]: Could you please change the check to `if (h2o_http2_stream_is_push(frame->stream_id))`? That is the wrapper function we have.

File: lib/http2/connection.c [hunk 2]
```diff
@@ -606,6 +606,10 @@ static int handle_priority_frame(h2o_http2_conn_t *conn, h2o_http2_frame_t *fram
         return H2O_HTTP2_ERROR_PROTOCOL;
     }
 
+    /* Ignore PRIORITY frames for pushed idle streams */
+    if (!(frame->stream_id & 1) && (frame->stream_id > conn->push_stream_ids.max_open))
+            return 0;
```
Comment (kazuho) [inline lib/http2/connection.c]: How about merging this with the `if` statement you added below?

The merits are:
* the if statements, combined would become simple, i.e.: `if (h2o_http2_stream_is_push(frame->stream_id)) { /* Ignore PRIORITY frames for pushed streams in idle or closed state */ return 0; } else { ... }`
* the check would move away from the hot path

**Comment by deweerdt** [comment]
@kazuho thank you for your comments, i believe they're addressed in bd95ec7

**Comment by kazuho** [comment]
Thank you for the changes!

**Pull request merged**